### PR TITLE
Add Tuya select

### PR DESCRIPTION
**Pull request by bearpawmaxim**
# What does this implement/fix?

Added Tuya select to use with enum datapoints type

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2075

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266

## Example entry for `config.yaml`:

```yaml
select:
  - platform: "tuya"
    tuya_id: tuyamcu
    name: "Warm floor sensor selection"
    enum_datapoint: 43
    options:
      0: Internal
      1: Floor
      2: Both

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/tuya/select/__init__.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        value = options_map_schema(value)
+
+        all_values = list(value.keys())
+        unique_values = set(value.keys())
+        if len(all_values) != len(unique_values):
+            raise cv.Invalid("Mapping values must be unique.")
+
+        return value
+
+    return validator
```
Comment (jesserockz) [inline esphome/components/tuya/select/__init__.py]: Don't really need the validator generator wrapper here
```suggestion
def ensure_option_map(value):
    cv.check_not_templatable(value)
    option = cv.All(cv.int_range(0, 2**8 - 1))
    mapping = cv.All(cv.string_strict)
    options_map_schema = cv.Schema({option: mapping})
    value = options_map_schema(value)

    all_values = list(value.keys())
    unique_values = set(value.keys())
    if len(all_values) != len(unique_values):
        raise cv.Invalid("Mapping values must be unique.")

    return value
```

File: esphome/components/tuya/select/__init__.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+
+CONFIG_SCHEMA = select.SELECT_SCHEMA.extend(
+    {
+        cv.GenerateID(): cv.declare_id(TuyaSelect),
+        cv.GenerateID(CONF_TUYA_ID): cv.use_id(Tuya),
+        cv.Required(CONF_ENUM_DATAPOINT): cv.uint8_t,
+        cv.Required(CONF_OPTIONS): ensure_option_map(),
```
Comment (jesserockz) [inline esphome/components/tuya/select/__init__.py]: If above is applied
```suggestion
        cv.Required(CONF_OPTIONS): ensure_option_map,
```

File: esphome/components/tuya/select/__init__.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        cv.Optional(CONF_OPTIMISTIC, default=False): cv.boolean,
+    }
+).extend(cv.COMPONENT_SCHEMA)
+
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await cg.register_component(var, config)
+    options_map = config[CONF_OPTIONS]
+    await select.register_select(var, config, options=list(options_map.values()))
```
Comment (jesserockz) [inline esphome/components/tuya/select/__init__.py]:
```suggestion
    options_map = config[CONF_OPTIONS]
    var = await select.new_select(config, options=list(options_map.values()))
    await cg.register_component(var, config)
```

File: esphome/components/tuya/select/tuya_select.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  void set_optimistic(bool optimistic) { this->optimistic_ = optimistic; }
+  void set_select_id(uint8_t select_id) { this->select_id_ = select_id; }
+  void set_select_mappings(std::vector<uint8_t> mappings) { this->mappings_ = std::move(mappings); }
+
+ protected:
+  void control(const std::string &value) override;
+
+  Tuya *parent_;
+  bool optimistic_ = false;
+  uint8_t select_id_{0};
```
Comment (jesserockz) [inline esphome/components/tuya/select/tuya_select.h]: Not required
```suggestion
  uint8_t select_id_;
```

File: esphome/components/tuya/select/tuya_select.cpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+void TuyaSelect::control(const std::string &value) {
+  if (this->optimistic_)
+    this->publish_state(value);
+
+  auto idx = this->index_of(value);
+  if (idx.has_value()) {
+    uint8_t mapping = this->mappings_.at(idx.value());
+    ESP_LOGV(TAG, "Setting %u datapoint value to %u:%s", this->select_id_, mapping, value.c_str());
+    this->parent_->set_enum_datapoint_value(this->select_id_, idx.value());
```
Comment (jesserockz) [inline esphome/components/tuya/select/tuya_select.cpp]: Shouldn't this be `mapping`?
```suggestion
    this->parent_->set_enum_datapoint_value(this->select_id_, mapping);
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
Nice. I just got a new Tuya device that actually needs this so I was going to implement it. Now I dont have to =)

A few minor suggestions, and one potential mistake maybe with `mapping` below?

Otherwise looks good. Thanks

File: esphome/components/tuya/select/tuya_select.cpp (hunk 5 above)
Comment (bearpawmaxim) [reply esphome/components/tuya/select/tuya_select.cpp]: Yes, this was a bug. Thanks! Fixed.

File: esphome/components/tuya/select/tuya_select.h (hunk 4 above)
Comment (bearpawmaxim) [reply esphome/components/tuya/select/tuya_select.h]: Agree, no need to set default value to zero for a uint.

File: esphome/components/tuya/select/__init__.py (hunk 3 above)
Comment (bearpawmaxim) [reply esphome/components/tuya/select/__init__.py]: Okay, simplified.

File: esphome/components/tuya/select/__init__.py (hunk 1 above)
Comment (bearpawmaxim) [reply esphome/components/tuya/select/__init__.py]: Agreed, removed.

**Comment by bearpawmaxim** [comment]
Hi @jesserockz , thanks for remarks in the review.
Just one question, can't we review also https://github.com/esphome/esphome/pull/3466 ?
Without that thing (and the one from https://github.com/ssieb/esphome/commit/0136ac6ebe488bc85360132c676d2bc58494f457), my thermostat isn't fully working

**Comment by jesserockz** [comment]
> Just one question, can't we review also https://github.com/esphome/esphome/pull/3466 ?

I get there eventually :smile:

**Review by jesserockz — APPROVED** [review]
Thanks

**Pull request merged**